### PR TITLE
TDL-19745 unittests and retrylogic for 500 exceptions

### PR DESCRIPTION
**Pull request by rdeshmukh15**
# Description of change

Covered below scenarios in Unit Test Cases :
- All helper functions with positive and negative test case scenarios
- Back-off is getting executed thrice or not.    
- Retry unit test cases if it working properly or not
- Test cases for all kinds of exceptions handled
- Validating the interval's value passed in config (Allowed values : HOUR, DAY, WEEK  and MONTH)
- Raising exception for invalid interval value.

Tap Changes:
- Added Retry for 500 errors

# Manual QA steps
 - 
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

**Comment by kethan1122** [comment]
@rdeshmukh15 can u add **Unit Tests** stage in circleCI config.yml.
`- run: 
          name: 'Unit Tests'
          command: |
            source /usr/local/share/virtualenvs/tap-dixa/bin/activate
            pip install coverage
            nosetests --with-coverage --cover-erase --cover-package=tap_dixa --cover-html-dir=htmlcov tests/unittests
            coverage html`

File: tests/unittests/test_backoff.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        client = Client(api_token="test")
+
+        try:
+            # Verifying if the custom exception 'DixaClient429Error' is raised on receiving status code 429
+            _ = client.get("https://test.com", "/test")
+        except DixaClient429Error: #as api_error:
+            pass
+
+            # Verifying the retry is happening thrice for the 429 exception
+        self.assertEquals(mocked_send.call_count, 3)
```
Comment (RushT007) [inline tests/unittests/test_backoff.py:40]: Observed that we have not implemented backoff and retry for 5XX in the tap, implement that and add respective unittests as well.

File: tests/unittests/test_config_interval.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    Allowed values in interval are Hour,Day, Week and Month
+    """
+    def test_invalid_value_of_interval(self):
+        try :
+            config = {'interval': "MONTHH"}
+            IncrementalStream.set_interval(self,value=config['interval'])
+            response = IncrementalStream.get_interval(self)
+        except InvalidInterval as e :
+            expected_error_message = "invalid interval provided"
+            assert str(e) == expected_error_message
```
[the remark this answers is not in the document]
Comment (rdeshmukh15) [reply tests/unittests/test_config_interval.py:19]: Done the changes as suggested.

File: tests/unittests/test_backoff.py (hunk 1 above)
Comment (rdeshmukh15) [reply tests/unittests/test_backoff.py:40]: Done

File: tests/unittests/test_httperrorhandling.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    @mock.patch("requests.Session.request", side_effect=lambda *_, **__: Mockresponse("", 503))
+    def test_503_error_custom_message(self, *args):
+        """
+        Unit test to check proper error message for 503 status code.
+        """
+        with self.assertRaises(exceptions.DixaClient5xxError):
+            try:
+                self.client_obj.get("https://test.com", "/test")
+            except exceptions.DixaClientError as _:
+                self.assertEqual(str(_), "Server Error")
```
Comment (RushT007) [inline tests/unittests/test_httperrorhandling.py:104]: Can we have descriptive error message for 5XX than `"Server Error"`?

**Comment by RushT007** [comment]
HTTPTimeout is not implemented, please implement it.

**Review by RushT007 — CHANGES_REQUESTED** [review]

File: tap_dixa/exceptions.py [hunk 4]
```diff
@@ -20,15 +20,15 @@ def __init__(self, message=None, response=None):
 class DixaClient5xxError(DixaClientError):
     pass
```
Comment (RushT007) [inline tap_dixa/exceptions.py:21]: Move DixaClient5xxError class to below DixaClient4xxError.

File: tests/unittests/test_httperrorhandling.py (hunk 3 above)
Comment (rdeshmukh15) [reply tests/unittests/test_httperrorhandling.py]: Made the changes as per suggestion.

File: tap_dixa/exceptions.py (hunk 4 above)
Comment (rdeshmukh15) [reply tap_dixa/exceptions.py]: Done

**Review by RushT007 — APPROVED** [review]

**Review by RushT007 — APPROVED** [review]

**Pull request merged**